### ISSUE TITLE
Do we need to set a session on product pages?

### ISSUE THREAD

**Issue by v18** [woocommerce/woocommerce-gateway-stripe#1248]
Right now, a session is set on product pages:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fade49f410ab83e296ed0ce24e509d5f39f11f4d/includes/payment-methods/class-wc-stripe-payment-request.php#L97

Is that needed? For caching purposes, it would be great if we could avoid setting a session.

reference: p72Wkx-2Mv-p2

**Comment by v18** [comment]
We need to set a session for product pages when displaying product request buttons.

**Comment by colmtroy** [comment]
@allendav it would be great if this was filterable and/or configurable. I couldn't use this plugin right now on some high traffic WC stores purely it force a cache bypass on our product pages. I imagine many other high traffic WC stores wanting to use this plugin and who don't need payment request buttons would love the option of removing the session on product pages.

**Comment by lkraav** [comment]
> it would be great if this was filterable and/or configurable. I couldn't use this plugin right now on some high traffic WC stores purely it force a cache bypass on our product pages. I imagine many other high traffic WC stores wanting to use this plugin and who don't need payment request buttons would love the option of removing the session on product pages.

If `WC_Stripe_Payment_Request::$_this` was refactored to `public`, it would enable @colmtroy to remove this session action like `remove_action( 'template_redirect', [ WC_Stripe_Payment_Request::$_this, 'set_session' ] )`.

Right now https://github.com/woocommerce/woocommerce-gateway-stripe/blob/fade49f410ab83e296ed0ce24e509d5f39f11f4d/includes/payment-methods/class-wc-stripe-payment-request.php#L1329 creates an orphan object that can't be referenced from anywhere even if we want to.

**Comment by colmtroy** [comment]
Just some more follow up on this.
Quick aside - I'm seeing quite a few deprecated notices related to `wc_stripe_hide_payment_request_on_product_page` when running 5.6.1 (see below).

If I've just selected that the Payment request button appears on the Cart e.g.
![](https://i.imgur.com/02Gp94W.png)

Why would the plugin need to set a session at all on the product page?

`[05-Oct-2021 14:39:49 Europe/Prague] PHP Deprecated:  wc_stripe_hide_payment_request_on_product_page is <strong>deprecated</strong> since version 5.5.0 with no alternative available. Please configure Payment Request Button locations through the Stripe plugin settings. in /var/www/snip/htdocs/wp-includes/functions.php on line 5596
[05-Oct-2021 14:39:49 Europe/Prague] PHP Deprecated:  wc_stripe_hide_payment_request_on_product_page is <strong>deprecated</strong> since version 5.5.0 with no alternative available. Please configure Payment Request Button locations through the Stripe plugin settings. in /var/www/snip/htdocs/wp-includes/functions.php on line 5596`

**Comment by colmtroy** [comment]
Polite bump on this issue - sessions are still being set on single product pages in 5.8.1 when payment request buttons (or express checkouts as they're now called) are disabled single product pages
![](https://i.imgur.com/O02UPMR.png)

This is a significant issue for high volume WooCommerce stores in that product pages cannot be delivered from cache. 
Any way this can be prioritized given the severity of it for high traffic stores?

**Comment by iMazed** [comment]
Also mentioned in 4649958-zen.

**Comment by yapofire** [comment]
Is there any update on this or a way around it?

Running WC 6.0 and Stripe Plugin 5.9

If you have express checkout turned off on product pages, then no session should be started by default on Product Pages

This is killing my google pagespeed score right now, as cloudflare etc will obviously skip any pages with an active session cookie

Any workaround code I can add?

**Comment by kovshenin** [comment]
Bumped into this problem today, ended up disabling the entire Express Checkout option as a workaround. I briefly looked over the codebase there and I really don't understand why the need to initiate a unique session, given that an anonymous user has merely visited a product page.

Perhaps we can do it on a redirect, or an XHR when the user actually clicks the Buy/Checkout button?

**Comment by nburch** [comment]
Yeah this is still a BIG issue. If you have a high-traffic site page cache is a _must_. As it is now, you essentially have to choose page cache _OR_ express payment (e.g. Apple Pay, etc.) but you can't have both. It's a real bummer and seems like an unnecessary compromise.

**Comment by stuartduff** [comment]
Mentioned in 5232103-zen

**Comment by essentique** [comment]
I see you closed #2392. Will you be tracking the fix here? Will it get fixed eventually? It's been years.

**Comment by rbgly** [comment]
6201318-zen

**Comment by wyter** [comment]
6547785-zen, they're also in discussion with devs via Hackerone #2073473

**Comment by nburch** [comment]
Any update on this? It's been 3+ years. Breaking page cache on **every product page** of a high traffic site is not really a viable option :( 

I would humbly suggest that this get a bump to **High Priority**

**Comment by essentique** [comment]
I outlined the solution for them years ago in #2392. They just need to make that oneliner change.

**Comment by isaumya** [comment]
Polite bump on this matter. Has there been any official resolution to this issue yet?

**Comment by stefpb** [comment]
I ended up to hot replace this line on every update:

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/15ef4d74c72db01ba70c32f369273e16459cecf1/includes/payment-methods/class-wc-stripe-payment-request.php#L713

I use only product_page-shortcodes, so this works for me: `sed -i "0,/wc_post_content_has_shortcode( 'product_page' )/s//false/" woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php`.

Bad hack, i know!

**Comment by mattf10** [comment]
Agree this is a major problem since it prevents caching on product pages.   It's also totally unnecessary unless express checkout is enabled on single product pages.

I just launched a new multi-language shop which means product pages are slow.   No caching = performance killer.

Eventually, I figured out how to add code to Nginx / Fast-cgi to allow caching only after the session cookie is set, but ideally, Stripe would fix this and session would not always be created on product pages.

**Comment by collette-tamez** [comment]
Bumping this again. We're moving into the holiday season so we're expecting a big bump and traffic with no way to cache product pages. Discouraging to see such a big issue with no movement in four years.

**Comment by colmtroy** [comment]
Delighted to see this one finally addressed @malithsen !

**Comment by stefpb** [comment]
> I ended up to hot replace this line on every update:
> 
> [woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/15ef4d74c72db01ba70c32f369273e16459cecf1/includes/payment-methods/class-wc-stripe-payment-request.php#L713)
> 
> Line 713 in [15ef4d7](/woocommerce/woocommerce-gateway-stripe/commit/15ef4d74c72db01ba70c32f369273e16459cecf1)
>  return is_product() || wc_post_content_has_shortcode( 'product_page' ); 
> 
> I use only product_page-shortcodes, so this works for me: `sed -i "0,/wc_post_content_has_shortcode( 'product_page' )/s//false/" woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php`.
> 
> Bad hack, i know!



In Version 10.4.0 the code moved. Now this is what i need to do on every build:

`RUN sed -i "0,/wc_post_content_has_shortcode( 'product_page' )/s//false/" /usr/src/wordpress/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-express-checkout-helper.php`

@isaumya

**Comment by isaumya** [comment]
Happy to see that the Stripe plugin has officially fixed this issue in v10.4.0

**Comment by daledupreez** [comment]
@stefpb, are you showing the express checkout payment methods on your product pages? I ask because this issue should have been addressed by https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3738, so I want to make sure we understand more about why you're still seeing the issue.

**Comment by stefpb** [comment]
True, you're right. The new version also solve the problem for me. Thank you very much!
